module-system/module-implementations/sov-nft-module/src/call.rs

```diff
@@ -0,0 +1,300 @@
+        nft.exit_if_not_owned(context)?;
+        nft.owner = to.clone();
+        self.nfts.set(&token_identifier, &nft, working_set);
+        Ok(CallResponse::default())
+    }
+
+    pub(crate) fn update_nft(
+        &self,
+        collection_name: &str,
+        token_id: u64,
+        token_uri: Option<String>,
+        frozen: Option<bool>,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<CallResponse> {
+        let (collection_address, _) =
+            self.get_collection_by_name(collection_name, context, working_set)?;
+        let token_identifier = NftIdentifier(token_id, collection_address);
+        let mut nft = self.get_nft_by_id(&token_identifier, working_set)?;
+        nft.exit_if_frozen()?;
+        if let Some(val) = frozen {
+            nft.frozen = val
+        };
+        if let Some(murl) = token_uri {
+            nft.token_uri = murl
+        };
+        self.nfts.set(&token_identifier, &nft, working_set);
+
+        Ok(CallResponse::default())
+    }
+
+    fn exit_if_collection_exists(
+        &self,
+        collection_name: &str,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<()> {
+        let creator = context.sender();
+        let ca = get_collection_address(collection_name, creator.as_ref());
+        ensure!(
+            self.collections.get(&ca, working_set).is_none(),
+            format!(
+                "Collection with name: {} already exists creator {}",
+                collection_name, creator
+            )
+        );
+        Ok(())
+    }
+
+    fn exit_if_nft_exists(
+        &self,
+        token_id: TokenId,
+        collection_address: &CollectionAddress<C>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<()> {
+        let msg = format!(
+            "Collection with address {} already exists",
+            collection_address.0
+        );
+        ensure!(
+            self.nfts
+                .get(
+                    &NftIdentifier(token_id, collection_address.clone()),
+                    working_set
+                )
+                .is_none(),
+            msg
+        );
+        Ok(())
+    }
+
+    fn get_collection_by_address(
+        &self,
+        collection_address: &CollectionAddress<C>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<Collection<C>> {
+        let c = self.collections.get(collection_address, working_set);
+        if let Some(collection) = c {
+            Ok(collection)
+        } else {
+            Err(anyhow!(
+                "Collection with address: {} does not exist",
+                collection_address.0
+            ))
+        }
+    }
+
+    fn get_collection_by_name(
+        &self,
+        collection_name: &str,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<(CollectionAddress<C>, Collection<C>)> {
```

Comment:
It seems like it would be pretty easy to introduce a bug where someone fetches a collection by address and then mutates it - effectively skipping the `is_owner` check. Should we introduce an extra type so that `get_collection_by_address` returns a wrapped `Collection` which can't be stored back in the tree - effectively making it immutable?